src/domain/relay/limit-addresses.mapper.ts

```diff
@@ -119,48 +119,102 @@ export class LimitAddressesMapper {
-    let execTransaction: { data: Hex; to: Hex; value: bigint };
-    // If transaction is an execTransaction
-    try {
-      const safeDecodedData = this.safeDecoder.decodeFunctionData({
-        data: args.data,
+    const execTransactionArgs = this.getExecTransactionArgs(args.data);
+    // Not a valid execTransaction call
+    if (!execTransactionArgs) {
+      return false;
+    }
+
+    // Only ERC-20 transfer to other party is valid
+    if (this.erc20Decoder.helpers.isTransfer(execTransactionArgs.data)) {
+      return this.isValidErc20Transfer({
+        to: args.to,
+        data: execTransactionArgs.data,
       });
-      if (safeDecodedData.functionName !== 'execTransaction') {
-        return false;
-      }
-      const [to, value, data] = safeDecodedData.args;
-      execTransaction = { to, value, data };
-    } catch (e) {
+    }
+
+    // Only ERC-20 transferFrom to other party is valid
+    if (this.erc20Decoder.helpers.isTransferFrom(execTransactionArgs.data)) {
+      return this.isValidErc20TransferFrom({
+        to: args.to,
+        data: execTransactionArgs.data,
+      });
+    }
+
+    // Other ERC-20 methods are valid if they are...
+
+    // Only transaction to other party is valid
+    const toSelf = execTransactionArgs.to === args.to;
+    if (!toSelf) {
+      return true;
+    }
+
+    // Only transaction with no value is valid
+    const hasValue = execTransactionArgs.value > BigInt(0);
+    if (hasValue) {
       return false;
     }
 
-    // If data of execTransaction is an ERC20 transfer
+    // Cancellations and Safe calls (e.g. owner management) are valid
+    const isCancellation = execTransactionArgs.data === '0x';
+    return isCancellation || this.safeDecoder.isCall(execTransactionArgs.data);
+  }
+
+  private getExecTransactionArgs(data: Hex): {
+    to: Hex;
+    value: bigint;
+    data: Hex;
+  } | null {
     try {
-      const erc20DecodedData = this.erc20Decoder.decodeFunctionData({
-        data: execTransaction.data,
+      const safeDecodedData = this.safeDecoder.decodeFunctionData({
+        data,
       });
-      // If the ERC20 transfer targets 'self' (the Safe), we consider it to be invalid
-      return (
-        erc20DecodedData.functionName === 'transfer' &&
-        erc20DecodedData.args[0] !== args.to
-      );
+
+      if (safeDecodedData.functionName !== 'execTransaction') {
+        return null;
+      }
+
+      return {
+        to: safeDecodedData.args[0],
+        value: safeDecodedData.args[1],
+        data: safeDecodedData.args[2],
+      };
     } catch {
-      // swallow exception if data is not an ERC20 transfer
+      return null;
     }
+  }
 
-    // If a transaction does not target 'self' consider it valid
-    if (args.to !== execTransaction.to) {
-      return true;
+  private isValidErc20Transfer(args: { to: string; data: Hex }): boolean {
+    const erc20DecodedData = this.erc20Decoder.decodeFunctionData({
+      data: args.data,
+    });
```

Comment:
`decodeFunctionData` can throw. but I think it should be ok to throw here. This function should be called given that we know that we are dealing with an ERC20 Transfer already. Throwing also creates visibility in case we have application bugs.

Reply:
I added a comment to keep note in edb9749fb6c722676f61148ad36909dfd53d9f79.